deb/openmediavault/usr/share/php/openmediavault/functions.inc

```diff
@@ -758,7 +758,7 @@ function binary_convert($number, $fromPrefix, $toPrefix, $decimalPlaces = 0): ?s
 		return FALSE;
 	// Convert number to string and ensure the correct number of decimal
 	// numbers is set (workaround for strange bcmul behaviour).
-	$number = number_format($number, 0, "", "");
+	$number = number_format($number, $decimalPlaces, ".", "");
```

Comment:
During testing, I had a problem with the conversion of numbers with decimal places. The numbers were always rounded up or down (1.5kB => 2024B), see the test. When troubleshooting, this function was the cause, as it converted 1.5 (float) to 2 (string).

A better solution would be to determine the decimal places of the transferred number in the function (number_format) and then transfer these to the function (I tried it, but it didn't work.). This is because the parameter $decimalPlaces is actually intended for formatting the return value.

In the bcdiv function, it would also be better to dispense with $decimalPlaces, as this can also lead to incorrect calculations due to rounding.

Reply:
You are testing `binary_convert` in your additional test, but did the changes to `binary_format`. Can you explain please.

Reply:
Are you sure? This is my Version:
`
function binary_format($number, $options = NULL) {
	$prefixes = [ "B", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB" ];
	$exp = 0;
	$maxExp = count($prefixes);
	$decimalPlaces = 2;
	$indexed = FALSE;

	// Process additional function options.
	if (is_array($options)) {
		if (isset($options['decimalPlaces']))
			$decimalPlaces = $options['decimalPlaces'];
		if (isset($options['indexed']))
			$indexed = $options['indexed'];
		if (isset($options['fromPrefix']))
			$exp = array_search($options['fromPrefix'], $prefixes);
		if (isset($options['maxPrefix']))
			$maxExp = array_search($options['maxPrefix'], $prefixes);
	}

	$number = number_format($number, 0, "", "");
	while ((-1 != bccomp($number, "1024")) && ($exp < $maxExp)) {
		$exp++;
		$number = bcdiv($number, "1024", $decimalPlaces);
	}

	$result = [
		"value" => floatval($number),
		"unit" => $prefixes[$exp]
	];
	if (FALSE === $indexed) {
		$result = sprintf("%s %s", $number, $prefixes[$exp]);
	}

	return $result;
}

function binary_convert($number, $fromPrefix, $toPrefix, $decimalPlaces = 0): ?string {
	$prefixes = [ "B", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB" ];
	$fromIndex = array_search($fromPrefix, $prefixes);
	$toIndex = array_search($toPrefix, $prefixes);
	if ((FALSE === $fromIndex) || (FALSE === $toIndex))
		return FALSE;
	// Convert number to string and ensure the correct number of decimal
	// numbers is set (workaround for strange bcmul behaviour).
	$number = number_format($number, $decimalPlaces, ".", "");
	// Convert the given number into the requested binary unit.
	$steps = ($toIndex > $fromIndex) ? $toIndex - $fromIndex :
	  $fromIndex - $toIndex;
	for ($i = 0; $i < $steps; $i++) {
		if ($fromIndex < $toIndex) {
			$number = bcdiv($number, "1024", $decimalPlaces);
		} else {
			$number = bcmul($number, "1024", $decimalPlaces);
		}
	}
	return $number;
}
`

Reply:
> During testing, I had a problem with the conversion of numbers with decimal places. The numbers were always rounded up or down (1.5kB => 2024B), see the test. When troubleshooting, this function was the cause, as it converted 1.5 (float) to 2 (string).

Any chance to have a dedicated unit test for that?

Reply:
> Are you sure? This is my Version:
> 
> ```
> function binary_format($number, $options = NULL) {
> 	$prefixes = [ "B", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB" ];
> 	$exp = 0;
> 	$maxExp = count($prefixes);
> 	$decimalPlaces = 2;
> 	$indexed = FALSE;
> 
> 	// Process additional function options.
> 	if (is_array($options)) {
> 		if (isset($options['decimalPlaces']))
> 			$decimalPlaces = $options['decimalPlaces'];
> 		if (isset($options['indexed']))
> 			$indexed = $options['indexed'];
> 		if (isset($options['fromPrefix']))
> 			$exp = array_search($options['fromPrefix'], $prefixes);
> 		if (isset($options['maxPrefix']))
> 			$maxExp = array_search($options['maxPrefix'], $prefixes);
> 	}
> 
> 	$number = number_format($number, 0, "", "");
> 	while ((-1 != bccomp($number, "1024")) && ($exp < $maxExp)) {
> 		$exp++;
> 		$number = bcdiv($number, "1024", $decimalPlaces);
> 	}
> 
> 	$result = [
> 		"value" => floatval($number),
> 		"unit" => $prefixes[$exp]
> 	];
> 	if (FALSE === $indexed) {
> 		$result = sprintf("%s %s", $number, $prefixes[$exp]);
> 	}
> 
> 	return $result;
> }
> 
> function binary_convert($number, $fromPrefix, $toPrefix, $decimalPlaces = 0): ?string {
> 	$prefixes = [ "B", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB" ];
> 	$fromIndex = array_search($fromPrefix, $prefixes);
> 	$toIndex = array_search($toPrefix, $prefixes);
> 	if ((FALSE === $fromIndex) || (FALSE === $toIndex))
> 		return FALSE;
> 	// Convert number to string and ensure the correct number of decimal
> 	// numbers is set (workaround for strange bcmul behaviour).
> 	$number = number_format($number, $decimalPlaces, ".", "");
> 	// Convert the given number into the requested binary unit.
> 	$steps = ($toIndex > $fromIndex) ? $toIndex - $fromIndex :
> 	  $fromIndex - $toIndex;
> 	for ($i = 0; $i < $steps; $i++) {
> 		if ($fromIndex < $toIndex) {
> 			$number = bcdiv($number, "1024", $decimalPlaces);
> 		} else {
> 			$number = bcmul($number, "1024", $decimalPlaces);
> 		}
> 	}
> 	return $number;
> }
> ```

You added the unit test https://github.com/openmediavault/openmediavault/pull/1853/files#diff-c0c9ee8cb634a1ff06471666c0c98504e0e68fe4d7019e7c33e12a6dc1b82cd6 which is testing `binary_convert` but did the changes to `binary_convert` with https://github.com/openmediavault/openmediavault/pull/1853/files#diff-a8129dd2522f02ec80e68be5708e3c86aa228886845095cb6534f0533588e811R761. I can not see any correlation to your comments here.

Reply:
already committed?
```
	public function test_binary_convert_4() {
		$str = binary_convert("1.5", "KiB", "B", 1);
		$this->assertEquals($str, "1536.0");
	}
```

Reply:
> You are testing `binary_convert` in your additional test, but did the changes to `binary_format`. Can you explain please.

I'm a little confused because you mention that I made the change in `binary_format`.

Reply:
> I'm a little confused because you mention that I made the change in `binary_format`.

The modification is part of your commit be9e2555dfe77388d3c721587e2727b7491eb18b.

Reply:
> It's included in your PR.

So is it wrong? Do you want a seperate commit/pull request with only a test that indicates the binary_convert bug?

Reply:
> > I'm a little confused because you mention that I made the change in `binary_format`.
> 
> The modification is part of your commit [be9e255](https://github.com/openmediavault/openmediavault/commit/be9e2555dfe77388d3c721587e2727b7491eb18b).

Yes and the change was made in `binary_convert`not `binary_format`.

![image](https://github.com/user-attachments/assets/f9b22a42-111e-4eee-a38c-841509ef4d39)